### PR TITLE
[Merged by Bors] - port: several files from category_theory

### PR DESCRIPTION
**Pull request by kim-em**
mathlib git sha: 8350c34a64b9bc3fc64335df8006bffcadc7baa6

Also ports the reassoc attribute.

- [x] depends on: #755

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#755
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Category/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+```lean
+local notation f ` ⊚ `:80 g:80 := category.comp g f    -- type as \oo
+```
+
+## Porting note
+I am experimenting with using the `aesop` tactic as a replacement for `tidy`.
+-/
+
+
+library_note "category_theory universes"/--
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Category/Basic.lean:35]:
```suggestion
library_note "category_theory universes"
/--
```

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#755~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Iso.lean [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+This file defines isomorphisms between objects of a category.
+
+## Main definitions
+
+- `structure Iso` : a bundled isomorphism between two objects of a category;
+- `class IsIso` : an unbundled version of `iso`;
+  note that `IsIso f` is a `Prop`, and only asserts the existence of an inverse.
+  Of course, this inverse is unique, so it doesn't cost us much to use choice to retrieve it.
+- `inv f`, for the inverse of a morphism with `[IsIso f]`
+- `as_iso` : convert from `IsIso` to `Iso` (noncomputable);
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Iso.lean:21]:
```suggestion
- `asIso` : convert from `IsIso` to `Iso` (noncomputable);
```

File: Mathlib/CategoryTheory/Functor/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+theorem map_dite (F : C ⥤ D) {X Y : C} {P : Prop} [Decidable P]
+    (f : P → (X ⟶ Y)) (g : ¬P → (X ⟶ Y)) :
+    F.map (if h : P then f h else g h) = if h : P then F.map (f h) else F.map (g h) := by aesop_cat
+#align category_theory.functor.map_dite CategoryTheory.Functor.map_dite
+
+-- Porting note: `to_prefunctor_obj` and `to_prefunctor_map` are now tautologies,
+-- so have not been ported.
+
+@[simp]
+theorem to_prefunctor_comp (F : C ⥤ D) (G : D ⥤ E) :
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Functor/Basic.lean:132]: I am wondering whether `to_prefunctor_comp` should be named `toPrefunctor_comp`? If so, there are a few other names which should be changed.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**